### ISSUE TITLE
Log output performance, rendering, possibly freezing

### ISSUE THREAD

**Issue by choffmeister** [drone/drone-ui#39]
I already mentioned this in my mass JS issue #15, and would like to work on this particular part now: The log output can get very long which makes viewing the logs taking possibly a long time before the DOM has been manipulated. There is a library called [react-virtualized](https://github.com/bvaughn/react-virtualized) which comes with some components to virtualize the rendering. Meaning you can tell him to evaluate in code which items of a big list are visible at all and only render these while still giving a scrollbar. On scrolling the new items in view get added to the DOM and the old get removed. This increases initial rendering speed as well as rendering speed while scrolling.

Interested in a PR regarding this?

**Comment by bradrydzewski** [comment]
how does `react-virtualized` handle items that do not have fixed height? The lines in the terminal will wrap and may span multiple lines when wrapping. We should expect the wrapping to change as the user resizes the browser width.

also wondering how it handles the fact that the list does not have a fixed height. I'm not terribly experienced in these areas but have seen virtual lists that require either fixed list height or fixed item height, which drone has neither without a decent amount of rework to the page layouts.

In general I think these optimizations are a good idea. But I will say that these are optimizations and I would love to direct effort at some of the current bugs and missing features, and then optimize.

**Comment by choffmeister** [comment]
In case of dynamic height due to ordinary text wrap this does not work by default (because to know the height of a line the browser would need to render the items, which is what should be prevented for most lines). So to work the height of individual items has to been known in code.

Possible solution:

Control wrapping from code rather then by browser text measurement. Set the maximal character count before wrapping in code (according to viewport width) and then calculate on the string how many lines are actually needed to display a single line. This is just some simple math. In the end we would have a list of items in memory with the additional knowledge of how many lines each line needs without having to render the lines before. Since the log output is monospaced font, this should work pretty well.

**Comment by bradrydzewski** [comment]
ok looks like bold doesn't impact the monospaced character width, at least on chrome on mac. We can also calculate height based on number of carriage returns. We'll probably need to understand the exact text-overflow strategy being used to make sure we can calculate correctly across browser engines.

**Comment by bradrydzewski** [comment]
another option could be to auto-collapse sections / commands as they finish, which should reduce the amount of overall dom items being displayed / rendered.

**Comment by choffmeister** [comment]
Yeah, auto collapsing is also a good idea. But anyway a single output can be arbitrarily large (see drone-ui npm install without --quiet :smile:), so the other thing should be done as well.

**Comment by choffmeister** [comment]
This is especially important on mobile devices.

**Comment by bradrydzewski** [comment]
@choffmeister I'm fully supportive of this change. I do think we'll need a diverse set of log samples and we'll need to test in multiple browsers to ensure we get the rendering right, if we end up having to do height calculations manually.

**Comment by choffmeister** [comment]
That is true. We should add a test suite before starting with this. Maybe I can do this on the weekend. Guess will find no time earlier for any further contribution.

**Comment by brgaulin** [comment]
@choffmeister @bradrydzewski Here is an example log file of our current build, but since its not a public project I went and fudged a good chunk of the data. If it is any better to get data from the DB let me know:
https://gist.github.com/brgaulin/c9eb3a9fca53199989f3585d258661e6

Today this is run via a shell script due to a bug in ancient docker. We are upgrading later this week, so ideally once 0.5 hits we could use the new pipeline setup to just put the commands right in there.

But we:
1. Compose: Redis, Memcached, MySQL
2. Pull from git
3. Bundle Install
4. Setup the DB
5. Precompile static assets for the tests
6. Run 4x parallel rspec integration tests
7. Run security static analysis
8. Run style static analysis
9. Bundle in deployment mode
10. Build static assets for production
11. Build a torquebox bundle
12. Ship the thing to artifactory

I would love to be able to collapse each of those pieces, so you could glance at the build and see why a build failed, and so we can find the error more quickly.

Some other things that would be awesome:
* Color the collapsed line when a step fails
* Ability to inject our own separators? For example, today we just run a rake task that does 99% of our work. Just one big happy command. I plan to break this apart, but could be cool to allow for a separator character to say start or end a command.

Some ideas that are a bit out there, but would be really cool:
* Ability to name the steps, so when they are collapsed it is easier to read
* As steps complete collapse the command
* Ability to attach files to the logging service, for example maybe the rails logs for debugging if tests failed
* Some kind of env variable that can be set to disable collapsing? I don't know why someone might want that, maybe you have a really short build. But maybe there should be a way to turn it on and off?

Not that copying someone else is ever a good idea, but I have really been liking how circle ci does their outputs
example: https://circleci.com/gh/InsightSoftwareConsortium/ITKMinimalPathExtraction/48

Sections with steps very clearly visible to what is going on, instead of feeling like you are looking at a wall of text.

**Comment by brgaulin** [comment]
Phew, with all that said, I'm curious, is there something I could help with? I can work on getting a local drone 0.5 setup at the very least so I can play with the output from our tests. But if I can help with testing or writing, please let me know. I'm not sure how far in you guys are already

**Comment by bradrydzewski** [comment]
@brgaulin thanks for the gist

I'm not sure if you've seen the latest UI but there have been many improvements to the output section. 0.5 stores structured logging data with line numbers, timestamps and more. The below screenshot represents some preliminary improvements with many more to come in future releases:

![screen shot 2016-07-17 at 7 25 00 pm](https://cloud.githubusercontent.com/assets/817538/16905002/718442e4-4c54-11e6-9212-2c11a865cd4d.png)


> Color the collapsed line when a step fails

we now track the exit code (appears at the bottom of the steps logs) so we should be able to surface this information when the step is collapsed. __Feel free to open an issue for this one__

> Ability to inject our own separators? For example, today we just run a rake task that does 99% of our work. Just one big happy command. I plan to break this apart, but could be cool to allow for a separator character to say start or end a command.

once command-level folding is implemented you can use `+ ` to signify command blocks, sort of like when you enable tracing in a bash program with `+x`. If you look at the 0.5 output you will see all commands are prefixed with `+`

> Ability to name the steps, so when they are collapsed it is easier to read

this is already done. Steps are named in your yaml and displayed in the UI

> As steps complete collapse the command

not implemented, but certainly possible. As mentioned above the last item in the logs for each step is the exit code, so it should be possible to react to this accordingly. __Feel free to open an issue for this one too__

> Ability to attach files to the logging service, for example maybe the rails logs for debugging if tests failed

the proposals to add artifacts or reporting capabilities to drone has been rejected. Instead we recommend 3rd party integrations (like coveralls). This is discussed below with suggested alternatives to storing reports and artifacts in drone:

https://github.com/drone/drone/issues/966#issuecomment-91316397
https://github.com/drone/drone/issues/239#issuecomment-80758786

> Some kind of env variable that can be set to disable collapsing? I don't know why someone might want that, maybe you have a really short build. But maybe there should be a way to turn it on and off?

I definitely see where you're coming from here. I have a general policy that we don't add features in anticipation of what people _might_ want. This prevents us from implementing a solution without full insight into the problem and real world usage patterns. It also keeps me focused on items we know will have a positive impact. But certainly interesting suggestion. Let's see what sort of feedback we get post-launch.

> Not that copying someone else is ever a good idea, but I have really been liking how circle ci does their outputs

I [discussed](https://discuss.drone.io/t/weekly-update-14-jul-2016/284/2) this in my latest update, actually. I would prefer to spend time creating an innovative build engine, as opposed to trying to innovate new and interesting user experiences. I therefore fully support implementing best practices that work well in other systems.

**Comment by subesokun** [comment]
> This is especially important on mobile devices.

It's even important on powerful devices :) I'd love to use Drone 0.5 in my team but currently the build logs are a blocker. We've here MacBook Pros but a simple build with a `npm install` causes already the browser to freeze for a few seconds and also scrolling feels quite slow (also with the latest UI changes). Especially when a build is running it's very challenging (almost impossible) to scroll around or perform other actions on the UI. As you've mentioned already this issue is not trivial so I'm looking forward to the solution you'll come up with :)

Maybe an easy fix for the initial log rendering issue of the builds that have been already completed could be to collapse every command that spans multiple lines into a single line and render only the first line of the command's output stream. I do believe that Travis does the same, so initial rendering would be very fast and if you're interested in the output of NPM then you can unfold it.

And maybe while the build is running the refresh rate of the logs array could be reduced so that an update of the list doesn't get triggered more frequently than the browser can render it. Normally that can be achieved via the `requestAnimationFrame` feature but probably React does optimize already such stuff...

**Comment by DracoBlue** [comment]
I worked around this issue in #128 to render only those logs, which are actually expanded!

**Comment by bradrydzewski** [comment]
solved in 1.0

**Comment by unfor19** [comment]
Hi @bradrydzewski
A question regarding the logging system (referring to drone-server version `1.9.1`) -

According to the docs - https://docs.drone.io/server/storage/blob/
> Drone stores large text files (e.g. build logs) in the database by default. Drone optionally supports storing large files in S3, or any object store compatible with the S3 protocol. This can significantly reduce the size of your database, and may improve overall performance and scalability.

Can I assume that ALL logs will be stored in S3, without truncating anything?

Expected behavior-
1. Truncates logs only in UI to avoid freeze - You wrote "solved in 1.0", so is this still relevant?
2. Stores all logs in S3 - Do I need to define a lifecycle policy in S3, or does Drone-server manage the lifecycle of the logs automatically?
